### PR TITLE
Rework troubleshooting section, add basic explainer

### PR DESCRIPTION
**Pull request by michaelpj**
This is an attempt to make the troubleshooting doc a bit better
structured. It's a bit lacking in concrete details for some of the
common problems (because I forget exactly how they mainfest), but I hope
that we can beef this up as things come up.

I also added a short explainer doc which explains some of the key
concepts.

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/2517"></a>

### CODE REVIEW

**Comment by michaelpj** [comment]
(reminder that there's a link to a preview of the rendered docs in the PR statuses)

**Review by jneira — APPROVED** [review]
Marvellous, i had a good time reading it and no everybody writes tech docs getting that.
Specially love the intro to basic concepts, a knowledge which we, advanced users, take worngly for granted

File: docs/what-is-hls.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+Here are a few pieces of jargon that you may come across in the HLS docs or when discussing problems:
+
+- *Code action*: A code action is a specific action triggered by a user on a particular region of code. Examples might include "add a type signature to this function".
+- *Code lens*: A pre-rendered edit or action shown in the body of the document itself, usually triggered with a click. Examples might include "the type signature for a function, which is actually inserted on click".
+- *Completion item*: An item that can be inserted into the text, including its metadata.
+- *Diagnostic*: Any information about the project that is shown in the editor, including errors, warnings, and hints from tools such as hlint.
+- *Semantic highlighting*: Special syntax highlighting performed by the server.
+
+## The structure of HLS
```
Comment (fendor) [inline docs/what-is-hls.md:37]: Oooh, is that a new section? Looks promising! Not sure whether you've seen it, but for GSoC 2020 bubba created a nice visualization: https://mpickering.github.io/ide/posts/2020-05-08-state-of-haskell-ide.html

Maybe we can create something similar and link it here?

File: docs/what-is-hls.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+Plugins can also be disabled independently to allow users to customize the behaviour of HLS to their liking.
+
+These plugins all (currently) live in the HLS repository and are developed in tandem with the core HLS functionality.
+
+### hie-bios
+
+HLS needs to know how to build your Haskell project: what flags to pass, what packages to provide, etc.
+It gets this information from the build system used by your project (typically `cabal` or `stack`).
+The tool used to do this is called [`hie-bios`](https://github.com/haskell/hie-bios).
+`hie-bios` calls the strategy it uses to get compilation flags (e.g. "ask `cabal`") a "cradle".
```
Comment (jneira) [inline docs/what-is-hls.md:58]: I would mention informally here what is a "cradle", we are using that word continuously and it is used in the new troubleshooting.

Also would be really great describe the implicit-hie discovery process (not sure if mentioning the library). But that can be left for a follow up

Comment (michaelpj) [reply docs/what-is-hls.md:58]: Yeah, although I'd welcome some contributions here. I'm not sure _I_ really know what goes on...

File: docs/what-is-hls.md (hunk 1 above)
Comment (michaelpj) [reply docs/what-is-hls.md:37]: I think it's probably misnamed :joy: It's more "here's some stuff we talk about when we talk about HLS that's useful to know about", like plugins! and hie-bios! It's not _really_ an explanation of the structure...

File: docs/troubleshooting.md [hunk 3]
```diff
@@ -1,79 +1,165 @@
...
+Sometimes HLS will simply fail to do anything with a file, or give nonsensical error messages.
+The most common cause of this is that HLS is using the wrong `hie-bios` cradle to decide how to build the project (e.g., trying to use `stack` instead of `cabal`).
+The server log will show which cradle is being chosen.
+
+Using an explicit `hie.yaml` to configure the cradle can resolve the problem, see the [configuration page](./configuration.md#configuring-your-project-build).
+
+### Problems with dynamic linking
+
+The prebuilt binaries which we provide are statically linked, which means that they don't work well with projects which use dynamic linking.
+A typical symptom is the presence of errors containing `unknown symbol`, and it is common for users Arch Linux, which ships a dynamically linked version of `ghc`.
```
Comment (pepeiborra) [inline docs/troubleshooting.md:134]:
```suggestion
The prebuilt binaries which we provide are statically linked, which means that they don't work well with projects which use dynamic linking.
```
What is the source for this?

The problems with statically linked HLS binaries are:

1. If `glibc` is statically linked, as in our Linux binaries, all sorts of unspecified behaviour can arise, including segfaults, when dynamically loading dependencies while compiling TH.
2. Because the GHC linker is used to dynamically load dependencies, and the GHC linker has bugs (mainly the Mach linker used in Mac OS, but also the ELF linker), HLS can segfault in projects with TH. 

2 means that HLS will crash more than `ghci`, as the latter uses the system linker instead of the GHC linker. Dynamically linked HLS binaries avoid this.

File: docs/what-is-hls.md (hunk 2 above)
Comment (jneira) [reply docs/what-is-hls.md:58]: hmm, err, well, what is really a cradle? @fendor? 😆

File: docs/troubleshooting.md (hunk 3 above)
Comment (michaelpj) [reply docs/troubleshooting.md:134]: The source is that it was already in the doc :) I don't understand the issue here so I'm happy to take a more accurate rewrite!

File: docs/troubleshooting.md [hunk 4]
```diff
@@ -1,79 +1,165 @@
...
+The workaround is to use a version of HLS compiled from source with the `ghc` option `-dynamic` enabled. See more details [here](https://github.com/haskell/haskell-language-server/issues/1160#issuecomment-756566273).
+
+### Support for Template Haskell
+
+Template Haskell should work fine in Linux and Windows with the distributed binaries.
+In MacOs a dynamically linked binary of HLS is required, otherwise use of TH can lead to segmentation faults.
+The easiest way to obtain a dynamically linked HLS binary is to build it locally.
+With `cabal` this can be done as follows:
+
+    cabal update && cabal install haskell-language-server --enable-executable-dynamic
```
Comment (pepeiborra) [inline docs/troubleshooting.md:145]:
```suggestion

### Template Haskell

**tl;dr** Use a dynamically linked HLS binary.

The easiest way to obtain a dynamically linked HLS binary is to build it locally. With `cabal` this can be done as follows:

    cabal update && cabal install haskell-language-server --enable-executable-dynamic

With `stack` you need to manually add the ghc option `-dynamic`.

Static binaries use the GHC linker for dynamically loading dependencies when typechecking TH code, and this can run into issues when loading shared objects linked against mismatching system libraries, or into GHC linker  bugs (mainly the Mach linker used in Mac OS, but also potentially the ELF linker). Dynamically linked binaries (including`ghci`) use the system linker instead of the GHC linker and avoid both issues. 

Note: HLS binaries prior to 1.6.0 were statically linking `glibc` which is not a supported configuration and has been replaced by `musl`.

**Comment by Anton-Latukha** [comment]
[skip circleci]

**Comment by Anton-Latukha** [comment]
Solves https://github.com/haskell/haskell-language-server/issues/491

File: docs/what-is-hls.md (hunk 1 above)
Comment (fendor) [reply docs/what-is-hls.md:37]: But it could be! I really loved https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/architecture.md which explains what each component does. However, what I have just linked is way more technical than our current section. But maybe some future work? Nevertheless, just wanted to say, I like the new section :D

**Comment by jneira** [comment]
@Anton-Latukha hmm how does this close #2491?, it needs the actual remove of the subproject and change the installation docs, which are not touched here

File: .github/ISSUE_TEMPLATE/bug_report.md [hunk 5]
```diff
@@ -8,17 +8,19 @@ assignees: ''
...
 -->
 
 ### Your environment
 
 Which OS do you use:
 <!-- Windows, MacOS, Ubuntu, ArchLinux, etc... -->
-Which lsp-client do you use:
-<!-- Neovim, emacs, VS Codium, etc... -->
+Which LSP client (editor/plugin) do you use:
+<!-- Neovim+LanguageClient-neovim, emacs+lsp-mode, VS Codium+vscode-haskell, etc... -->
```
Comment (Anrock) [inline .github/ISSUE_TEMPLATE/bug_report.md:23]: If I may: IIRC LanguageClient-neovim is deprecated long time ago and was incorporated into neovim itself. So majority modern neovim setups use either builtin lsp or coc.nvim.

File: docs/troubleshooting.md [hunk 6]
```diff
@@ -1,79 +1,165 @@
...
-## Common issues
+## Basic diagnosis steps
 
-### Problems with dynamic linking
+### Quick diagnosis checklist
 
-As haskell-language-server prebuilt binaries are statically linked, they don't play well with projects using dynamic linking.
-An usual symptom is the presence of errors containing `unknown symbol` and it is typical in arch linux, where a dynamically linked version of ghc is used.
+1. Is the server binary running at all? (check process monitors etc.) If not, there may be a client or installation problem.
+2. Is the client connecting to the server? (check for indicators in the editor etc.) If not, there may be a client problem.
```
Comment (Anrock) [inline docs/troubleshooting.md:27]: Should we explain this further since at least in neovim there is no indicators by default but there is a command that user has to execute manually?

File: docs/what-is-hls.md (hunk 1 above)
Comment (michaelpj) [reply docs/what-is-hls.md:37]: Future work :)

File: docs/troubleshooting.md (hunk 4 above)
Comment (michaelpj) [reply docs/troubleshooting.md]: Thank you!

File: .github/ISSUE_TEMPLATE/bug_report.md (hunk 5 above)
Comment (michaelpj) [reply .github/ISSUE_TEMPLATE/bug_report.md:23]: Thanks!

File: docs/troubleshooting.md (hunk 6 above)
Comment (michaelpj) [reply docs/troubleshooting.md:27]: I guess each of these probably needs a paragraph...

Comment (michaelpj) [reply docs/troubleshooting.md:27]: I might leave the first version light on detail, but I would very much welcome contributions from people who know e.g. the vim setup better. At least it would be nice to have a link to the vim client documentation in here.

Comment (Anrock) [reply docs/troubleshooting.md:27]: Ok. I guess I'll just wait until this PR is merged with whatever content you see fit and will create another one with vim details if necessary.

**Comment by Anton-Latukha** [comment]
Thank you @michaelpj for these updates, especially troubleshooting guidance for the community. It would be easier after this.
